e2e: expedited proposal test

czarcas7ic opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #2087 

## What is the purpose of the change

This pull request adds the ability to test if the newly introduced expedited proposals work as intended


## Brief Changelog

- Fixes small Makefile bug I introduced in a previous PR
- Updates SDK fork to 0.45.0x-osmo-v12-alpha.1 which includes [this PR](https://github.com/osmosis-labs/cosmos-sdk/pull/303)
- Adds `ExpeditedVotingPeriod` variable
- Adds `isExpedited` arg to TextProps and Deposits
- Adds `QueryPropStatus` function to query a proposal's status (passed, rejected, etc)
  - Adds a helper function `QueryPropStatusTimed` that time's how long it takes to get to a desired proposal status
- Fixes bug in [configurer/current.go](https://github.com/osmosis-labs/osmosis/pull/2341/files#diff-fc3948505a8cdd85079639e3a9d59393dce6c00a48e2a040522bd70028bcf607) that wouldn't take votingPeriod properly
- Comments out further changes [tests/e2e/containers/containers.go](https://github.com/osmosis-labs/osmosis/pull/2341/files#diff-5f119aa77c3b92f785aace7eab66b696056f43373dec9857f0ad466cdd58324e) that will be needed once v12 is introduced
- Adds `TestExpeditedProposals` that holds the ultimate expedited proposal logic


## Testing and Verifying

CI will NOT properly test this. As discussed with @p0mvn, we only run expedited proposal tests if upgrade tests are disabled until v12 is released due to compatibility issues. This can still be run locally though with upgrades disabled to show that expedited proposals work as intended!

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

czarcas7ic commented:
This should be R4R. It's unclear to me why gotest are failing. The only sdk diff between main and this are three commits, none of them having to do with mint 

https://github.com/osmosis-labs/cosmos-sdk/compare/0.45.0x-osmo-v11-alpha.5...v0.45.0x-osmo-v12-alpha.1

czarcas7ic commented:
Specifically it's this line here https://github.com/osmosis-labs/osmosis/blob/50b81179cb87dab5161c2e83e4ae67e5fb7e7a1d/x/mint/keeper/keeper_test.go#L354

czarcas7ic commented:
Ah I see what it is, we now return 0 for `GetSupplyWithOffset` instead of a negative number. I will change this test to reflect

alexanderbez commented on tests/e2e/configurer/chain/commands.go:
```diff
@@ -105,3 +114,26 @@ func (n *NodeConfig) CreateWallet(walletName string) string {
+			if err != nil {
+				return false
+			}
+			if status != desiredStatus {
+				return false
+			}
+			return true
```
```suggestion

			return status == desiredStatus
```

alexanderbez reviewed (approved):
Excellent work 👏

czarcas7ic commented:
This is blocked on a consensus in regards to this sdk PR: https://github.com/osmosis-labs/cosmos-sdk/pull/308

czarcas7ic commented:
@alexanderbez thank you!

@p0mvn addressed your comment from the sdk PR, agreed that setting the test offset to 1 is the best way. thanks for talking through that with me!

Now unblocked 😄

p0mvn commented on tests/e2e/configurer/current.go:
```diff
@@ -44,7 +45,16 @@ func (cb *CurrentBranchConfigurer) ConfigureChain(chainConfig *chain.Config) err
+
+	numVal := float32(len(chainConfig.ValidatorInitConfigs))
+
+	chainConfig.VotingPeriod = config.PropDepositBlocks + numVal*config.PropVoteBlocks + config.PropBufferBlocks
+	chainConfig.ExpeditedVotingPeriod = config.PropDepositBlocks + numVal*config.PropVoteBlocks + config.PropBufferBlocks - 1
+
```
Seems that we do a similar calculation in `upgrade.go`. Can we encapsulate this into a shared function or a method defined on the chain config?

p0mvn commented on tests/e2e/containers/containers.go:
```diff
@@ -212,8 +212,10 @@ func (m *Manager) RunNodeResource(chainId string, containerName, valCondifDir st
-func (m *Manager) RunChainInitResource(chainId string, chainVotingPeriod int, validatorConfigBytes []byte, mountDir string, forkHeight int) (*dockertest.Resource, error) {
+func (m *Manager) RunChainInitResource(chainId string, chainVotingPeriod, chainExpeditedVotingPeriod int, validatorConfigBytes []byte, mountDir string, forkHeight int) (*dockertest.Resource, error) {
 	votingPeriodDuration := time.Duration(chainVotingPeriod * 1000000000)
+	// TODO: Uncomment this after v12 release
```
Let's make an issue for this, please

czarcas7ic replied:
https://github.com/osmosis-labs/osmosis/issues/2386, I can likely take this post v12

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -211,3 +212,33 @@ func (s *IntegrationTestSuite) TestStateSync() {
+
+func (s *IntegrationTestSuite) TestExpeditedProposals() {
+	if !s.skipUpgrade {
+		s.T().Skip("this can be re-enabled post v12")
```
Let's make an issue and link it here with a comment please

czarcas7ic replied:
https://github.com/osmosis-labs/osmosis/issues/2386, included this in the same issue as above, will likely take this post v12 since I have context

p0mvn commented on x/mint/keeper/keeper_test.go:
```diff
@@ -336,6 +336,12 @@ func (suite *KeeperTestSuite) TestSetInitialSupplyOffsetDuringMigration() {
+			// this is because a negative supply offset will always return zero
+			// by setting this to the supply + 1, we ensure we are correctly calculating the offset by keeping it delta positive
+			supply := bankKeeper.GetSupply(ctx, sdk.DefaultBondDenom)
+			mintKeeper.MintCoins(ctx, sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, supply.Amount.Add(sdk.NewInt(1)))))
```
I think we should instead mint the amount that is equal to `keeper.DeveloperVestingAmount`. While these are equal with supply, it might be less apparent why we are using supply here. 

The ultimate reason for having the offset in the first place is having developer vesting rewards so using the amount related to it might be more obvious in terms of why this is done

czarcas7ic replied:
This is a really good catch, thank you!

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -211,3 +212,33 @@ func (s *IntegrationTestSuite) TestStateSync() {
+		s.T().Skip("this can be re-enabled post v12")
+	}
+
+	var wg sync.WaitGroup
```
What is the purpose of using both a channel and a wait group? Waitgroup seems redundant since we would block on a channel until it gives us a value anyways

czarcas7ic replied:
I actually wasn't aware that channels block until a value is returned. I tested it just to make sure and that is the case, thanks!

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -211,3 +212,33 @@ func (s *IntegrationTestSuite) TestStateSync() {
+	node.DepositProposal(chain.LatestProposalNumber, true)
+	wg.Add(1)
+	totalTime := make(chan time.Duration, 1)
+	go node.QueryPropStatusTimed(&wg, chain.LatestProposalNumber, "PROPOSAL_STATUS_PASSED", totalTime)
```
nit: to make it clearer
```suggestion
	go node.QueryPropStatusTimed(&wg, chain.LatestProposalNumber, "PROPOSAL_STATUS_PASSED", totalTimeChan)
```

p0mvn commented on tests/e2e/configurer/chain/commands.go:
```diff
@@ -105,3 +114,24 @@ func (n *NodeConfig) CreateWallet(walletName string) string {
 }
+
+func (n *NodeConfig) QueryPropStatusTimed(wg *sync.WaitGroup, proposalNumber int, desiredStatus string, totalTime chan time.Duration) {
+	defer wg.Done()
```
I think we can remove the waitgroup and instead send `1*time.Minute` on the `totalTime` channel since that's how long it takes at most for this query to wait. I think that would allow for the caller to get unblocked and continue. On the other hand, currently if `require.Eventually(...)` fails, the caller is blocked forever wiaitng on the result from `totalTime` channel that never arrives

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -211,3 +212,33 @@ func (s *IntegrationTestSuite) TestStateSync() {
 	s.NoError(err)
 }
+
+func (s *IntegrationTestSuite) TestExpeditedProposals() {
```
Is it possible to set the timeout specifically for this test case? Just in case it gets blocked due to multi-goroutine handling here

czarcas7ic replied:
There is a one minute timeout within the function the go routine is calling. Should I have a second timeout here as well? Would that be redundant or is it a good practice?

p0mvn replied:
Replied here: https://github.com/osmosis-labs/osmosis/pull/2341/files#r944758494

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -211,3 +212,33 @@ func (s *IntegrationTestSuite) TestStateSync() {
+	expeditedVotingPeriodDuration := time.Duration(chain.ExpeditedVotingPeriod * 1000000000)
+	timeDelta := elapsed - expeditedVotingPeriodDuration
+	// ensure delta is within one second of expected time
+	s.Require().Less(timeDelta, time.Second)
```
Can we have `time.Second` as a constant that we also use to set the parameter on chain? Just to make the link clear between how we configure it and the assertion here

czarcas7ic replied:
I don't think we would want this to be a set parameter right? I am basically just saying that the expeditedVotingPeriodDuration that this go routine calculates should realistically be within a second of the actual on chain param. I might be misunderstanding your recommendation here though

p0mvn replied:
Sorry my comment ended up being confusing.

What I meant was let's have a global constant defined for `expeditedVotingPeriodDuration`, `votingPeriodDuration`.

And also reuse those constants here:
https://github.com/osmosis-labs/osmosis/blob/6c65ea106b402245788f1390dbdfdbc9229fdfa6/tests/e2e/configurer/current.go#L48

Basically, the goal is to avoid hardcoding these values in different places. If we decide to change them in the future, we might update in one place but forget about another place and start having weird issues.

Having shared constants defined for such things helps mitigate these problems

p0mvn replied:
Actually, nvm I think that's already the case since we are using `chainConfig.ExpeditedVotingPeriod`

p0mvn replied:
The `*1000000000` part misled me. I'm going to resolve this, sorry for the confusion

p0mvn reviewed (commented):
Great work @czarcas7ic . I had some minor comments/thoughts but the solution looks great overall. Please let me know what you think

czarcas7ic reviewed (commented):
Thanks so much for the in depth review Roman! Only had a couple questions, everything else was clear!

p0mvn commented on tests/e2e/configurer/chain/commands.go:
```diff
@@ -105,3 +113,23 @@ func (n *NodeConfig) CreateWallet(walletName string) string {
+		"Osmosis node failed to retrieve prop tally",
+	)
+	elapsed := time.Since(start)
+	totalTime <- elapsed
```
This is related to an earlier comment:
https://github.com/osmosis-labs/osmosis/pull/2341/files#r944726827

Currently, the problem, is that we can fail after trying for 1 minute. However, nothin is sent on the `totalTime` channel if the failure does occur.

As a result, the test can get blocked for a long period of time (I think 30 minutes max)

I think we have 2 options to mitigate that and fail immediately:
1. Create a second timeout specifically for that test case (e.g. 2 minute timeout). So that if we block on the channel, the test case fails early

2. have a `defer` statement that either sends `elapsed` on the channel or a dummy value (i.e. -1)signifying failure.

Please let me know what you think

czarcas7ic replied:
I went back and forth between your two suggestion and decided to implement it through a case timeout (similar to the recommendation from Go by Example [here](https://gobyexample.com/timeouts))

I added a sleep to the timed go routine to ensure it stopped at the desired time and it did! I think this was the simplest implementation possible. Lmk what you think

p0mvn commented on tests/e2e/e2e_test.go:
```diff
@@ -214,3 +215,39 @@ func (s *IntegrationTestSuite) TestStateSync() {
+	case elapsed = <-totalTimeChan:
+	case <-time.After(timeoutPeriod):
+		err := fmt.Errorf("go routine took longer than %s", timeoutPeriod)
+		s.NoError(err)
```
```suggestion
		s.Require().NoError(err)
```

p0mvn reviewed (approved):
Nice! Great job
